s5/qlayers.py

```diff
@@ -4,6 +4,202 @@
 from .utils.quantization import q_dot_maybe, q_had_maybe, fully_quantized
 
 
+################### Extra imports for QLayerNorm ###################
+from typing import (Any, Callable, Iterable, Optional, Tuple, Union)
+from flax.linen.dtypes import canonicalize_dtype
+
+from flax.linen.module import Module, compact, merge_param  # pylint: disable=g-multiple-import
+from jax import lax
+from jax.nn import initializers
+import jax.numpy as jnp
+
+PRNGKey = Any
+Array = Any
+Shape = Tuple[int, ...]
+Dtype = Any  # this could be a real type?
+
+Axes = Union[int, Iterable[int]]
+#####################################################################
+
+
+def q_gelu(precision):
+    """
+        Quantized hard squish function to approximate GeLU.
+        Operates purely on integers without needing floating points.
+    """
+    _q_had = q_had_maybe(precision)
+
+    def _hard_sigmoid(x): # this operates purely on integers!
+        return jnp.minimum(jnp.maximum(0,x+2), 4) / 4 # jnp.right_shift allows for pure integer input/output!
+
+    def _q_gelu(x):
+        return _q_had(x, _hard_sigmoid(x))
+
+    return jax.jit(_q_gelu)
+
+
+
+def _compute_stats(x: Array, axes: Axes,
+                   dtype: Optional[Dtype],
+                   axis_name: Optional[str] = None,
+                   axis_index_groups: Any = None):
+  """Computes mean and variance statistics.
+
+  This implementation takes care of a few important details:
+  - Computes in float32 precision for stability in half precision training.
+  - mean and variance are computable in a single XLA fusion,
+    by using Var = E[|x|^2] - |E[x]|^2 instead of Var = E[|x - E[x]|^2]).
+  - Clips negative variances to zero which can happen due to
+    roundoff errors. This avoids downstream NaNs.
+  - Supports averaging across a parallel axis and subgroups of a parallel axis
+    with a single `lax.pmean` call to avoid latency.
+
+  Arguments:
+    x: Input array.
+    axes: The axes in ``x`` to compute mean and variance statistics for.
+    dtype: Optional dtype specifying the minimal precision. Statistics
+      are always at least float32 for stability (default: dtype of x).
+    axis_name: Optional name for the pmapped axis to compute mean over.
+    axis_index_groups: Optional axis indices.
+
+  Returns:
+    A pair ``(mean, var)``.
+  """
+  if dtype is None:
+    dtype = jnp.result_type(x) # TODO is this concerning???
```

Comment:
is this needed? why?

Reply:
They did this in the Flax source, I'm not sure if it's needed but the type promotion was worth flagging for further thought. I don't think it's an issue